src/models/UniqueNamedContainer.java

```diff
@@ -41,63 +43,54 @@ public UniqueNamedContainer() {
+     * There is another item with the same owner and the same original name.
+     * There is another item with a different owner and the same original name.
      *
-     * @param item Item to be added to the end of this container.
+     * @param item item to be added to the end of this container.
      */
     public void add(T item) {
         T newItem = (T) item.getCopy();
 
-        if (item.isGlobal()) {
-            // We only want to add global items it if its unique name is not present.
-            Optional<T> existing = findFirstWithUniqueName(newItem.getUniqueName());
-            if (existing.isEmpty()) {
-                items.add(newItem);
-            }
-        } else {
-            // Unique name naming rules:
-            //   Same owner and different name: keep it as is.
-            //   Different owner and name: keep it as is.
-            //   Different or the same owner and the same original name: add owner to name "owner.n.name" where n is a counter value.
-            // Motivation: Machine <= Machine || Machine. Here the owner is Machine but their clocks are different.
-            List<T> sameOriginalName = items.stream()
-                    .filter(current -> Objects.equals(current.getOriginalName(), newItem.getOriginalName()))
+        if (!item.isGlobal()) {
+            List<T> sameName = items.stream()
+                    .filter(it -> sameName(it, item))
                     .collect(Collectors.toList());
-
-            if (!sameOriginalName.isEmpty()) {
-
-                List<T> sameOwner = sameOriginalName.stream()
-                        .filter(current -> Objects.equals(current.getOwnerName(), newItem.getOwnerName()))
-                        .collect(Collectors.toList());
-
-                if (sameOwner.size() == 1) {
-                    sameOwner.get(0).setUniqueName(1);
-                }
-
-                if (!sameOwner.isEmpty()) {
+            if (sameName.size() != 0) {
+                List<T> sameOwner = sameName.stream().filter(c -> c.getOwnerName().equals(item.getOwnerName())).collect(Collectors.toList());
+                if (sameOwner.size() != 0) { // Same name, same owner
+                    for (int i = 0; i < sameOwner.size(); i++) {
+                        sameOwner.get(i).setUniqueName(i + 1);
+                    }
                     newItem.setUniqueName(sameOwner.size() + 1);
-                } else {
-                    for (T current : sameOriginalName) {
-                        current.setUniqueName();
+                } else { //  Same name, different owner
+                    for (int i = 0; i < sameName.size(); i++) {
+                        sameName.get(i).setUniqueName();
                     }
                     newItem.setUniqueName();
                 }
             }
+        }
 
+        // If the unique name is not present in the set of items then add it
+        Optional<T> existing = findFirstByUniqueName(newItem.getUniqueName());
```

Comment:
1) When the new item is `!isGlobal()`, this is effectively a redundant check, right? As the renaming done above should render the name of the new item unique.
2) For `isGlobal()` items, no renaming is performed. But now the caller has no idea whether a global item has been successfully added to the list. Maybe this method should return a boolean indicating whether the new item has been added or not? (See https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#add-E-)

Reply:
1. If the new item is `!isGlobal()` then we have to change its unique name, but when it is global we should not. I might be missing something, but I am pretty sure that the check is not redundant.
2. That is a very good idea.

Reply:
No. ``existing.isEmpty()`` should always be true for all "local" items. I have changed the code to reflect this.